net: gptp: Experimental Grand Master support

tgorochowik opened the pull request:
So far, the gPTP support in Zephyr only covered the Slave side. Zephyr could synchronize its clock to other devices but there was no way to for example sync clocks between two devices running Zephyr. This PR adds that possibility.
It is marked as experimental because the whole gPTP support still is.

What is added is basically what one of the commits says: GMCAP-1, GMCAP-2, and GMCAP-3  from 802.1AS-2011 are either implemented from scratch or finished. Additionally there are some commits with bug-fixes that were found along the way and are needed for the Grand Master role to work.

Fully working test setup, on which the GM support was tested:
`SAM e70 Xplained <-> regular Ethernet switch <-> SAM e70 Xplained`

The switch in between is useful to be able to also connect a third party host to it for sniffing (e.g. with ptpTrackHound), but it also makes the connections and bring up much easier, because sam gmac does not yet support link re-negotiation (it only negotiates link during boot), so it makes sense to have a device that is always on on the other side.

Some notes and known limitations:
* The correction field is always set to 0 in the Sync Follow Up packets - this mimics the behavior of OpenAvnu (a Linux gPTP client). There is a TODO left in the code which indicates that the filed should be properly calculated in order be fully compliant.
* OpenAvnu accepts Zephyr as GM but refuses to sync its clock - this could also be the case with some other gPTP-enabled devices.
* The clock accuracy can be chosen through KConfig - generally the chosen setting should reflect the capabilities of the HW, but this option is also useful for testing, as the device with the best accuracy is chosen as GM
* There is a `gptp_clock_source_time_invoke` callback added. According to the standard it should be called every time the clock updates its value but an `ON_DEMAND` functionality was added to probe the clock from current master when it is actually needed, as it doesn't always make sense to update it continuously in the ptp clock driver.

tgorochowik committed:
net: gptp: Rename the ClockMasterSyncReceive state struct

There are other ClockMasterSync state-machines (to be implemented).
Current name would either cause conflicts or be too ambiguous.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik committed:
net: gptp: Convert SyncReceiptTime to ExtendedTimestamp

SyncReceiptTime should use an ExtendedTimestamp (with fractional
nanoseconds precision). Add a struct with the definition of the needed
type and convert that variable.

The struct representing the ExtendedTimestamp is named
net_ptp_extended_time to keep consistency with the existing net_ptp_time
which is used for regular PTP timestamps.

See 802.1AS-2011 chapters 10.2.3.4 and 6.3.3.5 for more reference.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik committed:
net: gptp: Fix port number in SYN and FUP packets

Make sure the packets contain the port number they are actually coming
from.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik committed:
net: gptp: Initial GrandMaster capability support

Implements GMCAP-1, GMCAP-2, and GMCAP-3 and their dependencies from
802.1AS-2011. See Annex A.10 for more details.

The Grand Master Capability can be turned on and off through KConfig.

Note: the correction field in FUP packets is not yet properly
calculated. There is a TODO left in the code, near which some parameters
are zeroed to make the correction field be set to 0. This mimics the
behavior of openAvnu (a Linux gPTP client). For full compliance the
field should be calculated and set properly.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik committed:
net: gptp: Add clock accuracy selection

This commits adds a possibility to select PTP clock accuracy through
KConfig.

The chosen accuracy should reflect the capabilities of the used
hardware.

See IEEE 1588-2008, chapter 7.6.2.5 for more details.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik committed:
net: gptp: Fix sync receipt timeout timer starting

Completely remove the last_sync_receipt_timeout time. It is not part of
the standard (see 802.1AS-2011, 10.2.11 for the complete list of
variables for this state machine). Additionally this extra variable was
never really initialized so the calculated duration made no sense.

Just start the timer based on the regular sync receipt timeout time
interval from the port data set.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

tgorochowik committed:
net: gptp: Fix sync interval timeout timer starting

The last parameter is 'period', we should not pass duration twice.
Doing so makes zephyr hang with current mainline when it is chosen to be
the Grand Master.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

jukkar commented on include/net/gptp.h:
```diff
@@ -56,6 +56,49 @@ struct gptp_uscaled_ns {
+/**
+ * @brief Extended Timestamp
+ */
+struct gptp_extended_ts {
```
We have the net_ptp_time defined in include/net/ptp_time.h, what about creating `struct net_ptp_extended_time` in that file instead?

tgorochowik replied:
Changed

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_mi.c:
```diff
@@ -772,47 +793,331 @@ static void gptp_mi_clk_slave_sync_state_machine(void)
+	       GPTP_CLOCK_ID_LEN);
+
+	sync_info->src_port_id.port_number = 0;
+
```
Probably the empty lines here and few lines below could be removed when we set the fields of the same variable. Now it looks excessively spacy.

tgorochowik replied:
That's right, changed

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_state.h:
```diff
@@ -114,6 +114,18 @@ enum gptp_pa_transmit_states {
+enum gptp_cms_offset_states {
+	GPTP_CMS_OFFSET_INITIALIZING,
+	GPTP_CMS_OFFSET_INDICATION,
+};
```
We could make this __packed to save some space in structs where this is used.

tgorochowik replied:
Sure, that makes perfect sense! I have a question though: shouldn't this be done more globally? Other enums are not `__packed` at the moment (at least not in the gptp layers). What is different about these particular enums? As far as I can see most of them could be packed to 1 byte. Maybe we should pack them all, or even suggest something like `-fshort-enums` to be used globally?

jukkar replied:
I think it makes only sense to use __packed here if we are using the enum inside a struct where size matters. See also this https://oroboro.com/short-enum/ for details.

tgorochowik replied:
Ok, thanks! I thought the `-fshort-enums` could be a little bit too far-fetched. Thanks for the explaination!

What about other enums for states for other state machines? For example `enum gptp_pa_rcv_states` and the corresponding `struct gptp_port_announce_receive`? Size-wise it looks exactly the same as `enum gptp_cms_offset_states` and its `struct gptp_clk_master_sync_offset_sate`. Is there something that I am missing?

jukkar replied:
Those enum's that are used inside structs could be made __packed. I did some conversion earlier for these in other parts of the net stack but it seems that I missed the gptp stuff. Could you add packed to these in this PR too?

tgorochowik replied:
Ok, sure! Thanks for explanation

tgorochowik replied:
Added a separate commit before the GM commit that adds this to the previously existing enums. I also moved some of the fields in the structs as after adding that attribute new holes were created in them.

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_state.h:
```diff
@@ -114,6 +114,18 @@ enum gptp_pa_transmit_states {
+enum gptp_cms_snd_states {
+	GPTP_CMS_SND_INITIALIZING,
+	GPTP_CMS_SND_INDICATION,
+};
```
same here

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_state.h:
```diff
@@ -374,6 +386,27 @@ struct gptp_clk_slave_sync_state {
+
+/* ClockMasterSyncSend state machine variables. */
+struct gptp_clk_master_sync_snd_state {
+	/** Current state of the state machine. */
```
Shuffle the struct fields so that there are no holes, for example the enum should be last here.

tgorochowik replied:
Done

jukkar commented on subsys/net/l2/ethernet/gptp/Kconfig:
```diff
@@ -21,6 +21,92 @@ config NET_DEBUG_GPTP
+
+choice
+	prompt "gPTP Clock Accuracy"
+	depends on NET_GPTP
```
Should we set the default value to unknown? The unknown field is the first one but it still it would probably make sense to say it here too.

tgorochowik replied:
Default value added

jukkar commented on include/net/gptp.h:
```diff
@@ -27,11 +27,10 @@
-#define GPTP_CLOCK_ACCURACY_UNKNOWN        0xFE
-
-#define GPTP_OFFSET_SCALED_LOG_VAR_UNKNOWN 0x436A
+#define GPTP_OFFSET_SCALED_LOG_VAR_UNKNOWN 0x6A43
```
Perhaps comment here explaining what this value is and in what endianess.

tgorochowik replied:
That is actually a very good question. The value was switched based on observations of the packets that openAvnu sends (as it made sense to mimic it in the beginning as it is something that is _known to work_) - unfortunatelly I cannot find any references to that value in 802.1AS - I will test it all again without this change, and if it still works I guess we can temporarily drop this patch until the question is properly answered.

jukkar replied:
Perhaps trying to find out the value handling from OpenAvnu sources could help here. There is lot of C++ code there and it is not easy to spot some specific place, I have tried something like this earlier.

tgorochowik replied:
Unfortunately the value is hard-coded there too (and without any comments):
https://github.com/AVnu/gptp/blob/98e3adaeb4c5a9534d24aee166d3690ea6a5c7f4/common/ieee1588clock.cpp#L95

The definition is the same as it was here before this patch, but Zephyr does `htons` on this value before sending this packet, openAvnu sends it as it is, so we needed to switch it to make it the same.

I will let you know how the tests without this patch go, I will also try to go through the standards to see if there are some details and suggested default values for this field.

jukkar replied:
Ok, that is tricky then. I wonder if linux-ptp daemon has anything better documented in its code.

tgorochowik replied:
I removed the patch for now as this value doesn't seem to affect sync process.

tgorochowik replied:
But for the sake of completeness:
* linux-ptp sets this field to `0xffff` by default
* 802.1as-2001, 8.6.2.4b says, that if `offsetScaledLogVariance` is not specified or not know, then this field should be set to `0x4100` - maybe that is the way to go? None of the daemons we used use this value, but it seems to be the best choice actually (although since it doesn't seem to have any effect on sync then it is kind of low priority)

tgorochowik committed:
net: gptp: Use packed enums when applicable

Optimize memory usage of enums that are used within structs.
Reorganize the affected structs to avoid holes.

Signed-off-by: Tomasz Gorochowik <tgorochowik@antmicro.com>

jukkar commented on include/net/gptp.h:
```diff
@@ -207,6 +206,25 @@ struct gptp_phase_dis_cb {
+ *
+ * Parameters passed by ClockSourceTime.invoke function.
+ */
+struct gptp_clk_src_time_invoke_params {
```
This struct looks like there could be holes in it. You could check that with `pahole -C gptp_clk_src_time_invoke_params zephyr.elf` command. In fedora the pahole program can be found in dwarves package.

tgorochowik replied:
I actually used `pahole -P zephyr.elf` - it is supposed to find all structs with holes that can actually be packed. I assumed it is all that can be done, but I will try doing it the way you described, thanks!

jukkar replied:
ok, no issues then if you already paholed it :)

jukkar commented on include/net/gptp.h:
```diff
@@ -278,6 +296,14 @@ void gptp_foreach_port(gptp_port_cb_t cb, void *user_data);
+ *
+ * @param arg Current state and parameters of the ClockSource entity.
+ */
+void gptp_clock_source_time_invoke(struct gptp_clk_src_time_invoke_params arg);
```
Shouldn't the parameter be a pointer here?

tgorochowik replied:
Sure, it should, will change that, thanks!

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_mi.c:
```diff
@@ -771,47 +791,329 @@ static void gptp_mi_clk_slave_sync_state_machine(void)
+
+static void gptp_compute_gm_rate_ratio(void)
+{
+	static struct net_ptp_extended_time src_time_0 = {};
```
No need to initialize the static vars, they are set to 0 by default anyway.

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_data_set.h:
```diff
@@ -244,6 +244,12 @@ struct gptp_global_ds {
 
 	/** A Grand Master is present in the domain. */
 	bool gm_present;
+
```
Please put the structs after the other structs few lines above (trying to avoid extra holes here).

jukkar replied:
If pahole does not say there are holes in this struct, then I am fine with current placement.

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_user_api.c:
```diff
@@ -81,3 +81,15 @@ char *gptp_sprint_clock_id(const u8_t *clk_id, char *output, size_t output_len)
 	return net_sprint_ll_addr_buf(clk_id, 8, output, output_len);
 }
+
+void gptp_clock_source_time_invoke(struct gptp_clk_src_time_invoke_params arg)
```
Shouldn't the parameter be a pointer?

jukkar reviewed (approved):
LGTM

jinxdnatong commented:
I did experiments on stm32f407, and the time error of master and slave was 4ms. I checked the debug log, and slave always set the time ,this  because of the error of GPTP or my code error？

jukkar commented:
What kind of hw setup you have? I am asking because the current upstream stm32 ethernet driver does not have gPTP support.

jinxdnatong commented:
I add some gptp support on stm32 ethernet driver, but when i run zhe gptp,error time of master and slave is 4 ms, i don't know  
it because I added drivers, or gptp have bug，and I'm using the latest code on GitHub that supports master

jinxdnatong commented:
I add some gptp support on stm32 ethernet driver, but when i run zhe gptp,error time of master and slave is 4 ms, i don't know  
it because I added drivers, or gptp have bug，and I'm using the latest code on GitHub that supports master

jukkar commented:
It is difficult to say where is the problem in this case as you have out-of-tree driver updates. Would you be able to send your stm32 eth driver changes to upstream for a review?

jukkar commented:
You can submit a PR that has your changes to the stm32 eth driver. After that people are able to review your code and eventually the PR can be merged to upstream.
